src/storage/src/storage_state/async_storage_worker.rs

```diff
@@ -0,0 +1,166 @@
+        let (response_tx, response_rx) = crossbeam_channel::unbounded();
+
+        mz_ore::task::spawn(|| "AsyncStorageWorker", async move {
+            while let Some(cmd) = command_rx.recv().await {
+                // Peel off all available commands. This allows us to catch up
+                // if we fall behind.
+                let mut commands = vec![cmd];
+                while let Ok(cmd) = command_rx.try_recv() {
+                    commands.push(cmd);
+                }
+
+                for (span, command) in commands {
+                    match command {
+                        AsyncStorageWorkerCommand::CalculateResumeFrontier(
+                            id,
+                            ingestion_description,
+                            _phantom_data,
+                        ) => {
+                            let mut persist_clients = persist_clients.lock().await;
+                            let mut state = ingestion_description
+                                .initialize_state(&mut persist_clients)
+                                .instrument(span.clone())
+                                .await;
+                            let resume_upper: Antichain<T> = ingestion_description
+                                .calculate_resumption_frontier(&mut state)
+                                .instrument(span)
+                                .await;
+                            let res = response_tx.send(
+                                AsyncStorageWorkerResponse::IngestDescriptionWithResumeUpper(
+                                    id,
+                                    ingestion_description,
+                                    resume_upper,
+                                ),
+                            );
+                            activatable.activate();
+
+                            match res {
+                                Ok(_) => {
+                                    // All's well!
+                                }
+                                Err(_err) => {
+                                    // Receiver must have hung up.
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            tracing::trace!("shutting down async storage worker task");
+        });
+
+        Self {
+            tx: command_tx,
+            rx: response_rx,
+        }
+    }
+
+    /// Calculates a recent resume upper for the given `IngestionDescription`.
+    pub fn calculate_resume_upper(
+        &self,
+        id: GlobalId,
+        ingestion: IngestionDescription<CollectionMetadata>,
+    ) {
+        self.send(AsyncStorageWorkerCommand::CalculateResumeFrontier(
+            id,
+            ingestion,
+            PhantomData,
+        ))
+    }
+
+    fn send(&self, cmd: AsyncStorageWorkerCommand<T>) {
+        self.tx
+            .send((tracing::Span::current(), cmd))
```

Comment:
interesting, i would have forgotten to add the span here, what broke?

Reply:
Nothing! But I also wrote the persist workers in the storage controller, and they do a similar thing. You think it's not needed?

Reply:
No i think its a great addition!